exercises/practice/list-ops/.meta/proof.ci.wren

```diff
@@ -15,13 +15,21 @@ class ListOps {
 
   // https://wren.io/control-flow.html#the-iterator-protocol
   iterate(iterator) {
+    // first iteration, wren provides the null value
     if (iterator == null) {
-      if (_data.isEmpty) return false
-      return 0
+      if (_data.isEmpty) {
+        return false
+      } else {
+        return 0
+      }
     } else {
+      // subsequent iterations, wren provides the previous value
       iterator = iterator + 1
-      if (iterator == _data.count) return false
```

Comment:
Ditto.

Reply:
I'd probably do it with the whole function if/else as well though. :)

Reply:
Yes, this looks nice:
```wren
  iterate(iterator) {
    // first iteration, wren provides the null value
    if (iterator == null && _data.isEmpty) return false
    if (iterator == null) return 0
    // subsequent iterations, wren provides the previous value
    return (iterator + 1 == _data.count) ? false : iterator + 1
  }
```

Reply:
No.  I don't see any problem with how it was before... guard clause for the "Exceptional" case, then normal code path for the 'happy path'...

Reply:
For example I think this is quite nice:

```js
    // first iteration, wren provides the null value
    if (iterator == null) { // guard
      if (_data.isEmpty) return false // guard

      return 0
   }

   // happy path here
````

Reply:
My only objection is the wording of what you're saying: there aren't truly any exceptional cases: there are simply 4 branches through the happy path:
```
{is this the first iteration} × {are we at the end of the collection}
```

Reply:
> there aren't truly any exceptional cases: 

I **think** I _may_ understand what you mean, but to me the "before iterations starts" is definitely the outside case here... in a sequence of 100,000 things it will run once, the "happy path" 99,999 times. 1 out of x where x grows to infinity, is that not the exception?

Seriously would love to know more about how you think about this if you differ.  I mean I wouldn't have added explicit "happy path here" comments, that was for your sake (and it is how i think about it I guess)... if one branch of an if/else is called once while the other is called 100,000 times I would likely _never_ structure it as an if/else, I think that's "deceptive" to a reader and should imply equal (or close to equal execution paths)...